widget_driver/doc/widget_drivers.md

```diff
@@ -0,0 +1,142 @@
+        _localization = context.read<Localization>(),
+        super(context) {
+
+    _subscription = _counterService.valueStream.listen((_) {
+      notifyWidget();
+    });
+
+  }
+
+  @DriverProperty('The title of the counter')
+  String get counterTitle => _localization.counterTitle;
+
+  @DriverProperty(1)
+  int get value => _counterService.value;
+
+  @DriverAction()
+  void increment() {
+    _counterService.increment();
+  }
+
+  @DriverAction()
+  void decrement() {
+    _counterService.decrement();
+  }
+
+  @override
+  void dispose() {
+    _subscription?.cancel();
+    super.dispose();
+  }
+}
+```
+
+There are three important concepts here.
+
+1. The dependencies are loaded/crated/resolved in the constructor of the `driver`.
+1. All the properties/methods which the driver wants to expose to the widgets needs to be annotated and given default values.
+1. Whenever important data has changed in the `driver` you need to call `notifyWidget()`
+
+**Let's start talking about the dependencies.**  
+Any type of dependencies which your `driver` needs to be able to give your widget the correct data has to be loaded/crated/resolved in the constructor of the `driver`.
+
+You get two option here for how to resolve these dependencies. Either you can grab them out of the `BuildContext` (for example if you are using the `Provider` package). Or you can use some DI package like `get_it` and then just grab the dependencies from the DI container.
+
+**Now what about those annotations?**  
+They are needed for the `testDrivers` to be generated correctly. So for all properties and methods which you expose to the widget, you will need to add these annotations.
+
+For properties you add `@DriverProperty({default_value})`.  
+The `default_value` should be the default value which you want to use when this widget is being created by other widgets under test.
+
+For methods you add `@DriverAction()`.  
+This is needed so that any potential parent widget when tested can call these methods without crashing.
+
+After you are done creating your `driver` you need to run the generator so that the `testDrivers` gets created. Read more [here](code_generation.md) about running the generators.
+
+**Updating the widget.**  
+So now you have your dependencies and they have emitted some new state and you want to update your widget. But how?!
+
+Easy, you just call `notifyWidget()`. This will force your widget to call its `build` method again and update. Then it will just read the latest values from your `driver`.
+
+So for this to work, make sure that you first have all the correct data in place in your `driver` before calling `notifyWidget()`. Because as soon as you call `notifyWidget()` the widget will get updated.
+
+### Life cycle
+
+Your `driver` lives together with the widgets which it drives. As soon as that widget gets put on screen, the `driver` gets created. If that widget gets remove from the screen then your `driver` gets disposed.
```

Comment:
```suggestion
Your `driver` lives together with the widget it drives. As soon as that widget gets put on screen, the `driver` gets created. If that widget gets remove from the screen then your `driver` gets disposed.
```
In theory a driver could be used for multiple widget instances, right? But anyhow I'd stay singular (widget) here.

Reply:
Well not really. Each DrivableWidget gets its own instance of the driver... So all DrivableWidgets of the same type will share drivers of the same type.. But they all have their own instance of that driver..

So each widget has its own instance of the driver..

You should not/cannot share the same driver instance between multiple widgets.. Since the framework creates the instance for you.. So it is not really up to you..


But what you could do in theory is to share the same driver type between different drivableWidget type.. So you could have 
```
class MyFirstDrivableWidget extends DrivableWidget<MyDriver> {}

class MySecondDrivableWidget extends DrivableWidget<MyDriver> {}
```

In this case, both these widgets would have drivers of the same type..
But I would not recommend this.. Since the drivers are view models, they should be tailored for a unique view.. If you start sharing drivers, then you start putting shared business logic in them and in the end you end up with DigitalKeyBloc or VehiclesBloc..